### ISSUE TITLE
How to free XMLElement

### ISSUE THREAD

**Issue by robertfeldt** [JuliaIO/LightXML.jl#17]
I get mem leaks in a use of LightXML that creates a large number of deeply nested XMLElements. I tried to add a free method and use it as finalizer, see these diffs/commits:

https://github.com/robertfeldt/LightXML.jl/commit/238fbb5c41629e254088dfe7553a53c72bd5dffa
https://github.com/robertfeldt/LightXML.jl/commit/62d4a7a484915642de78bd0d6e8552269364e47a
https://github.com/robertfeldt/LightXML.jl/commit/43833a5dd0dfc0c258d5d7797fde8b158b979824

but when I run I get:

```
julia(56180,0x7fff7924c300) malloc: *** error for object 0x7fc8d9efbef0: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
```

I appreciate any help on how one can properly free XMLElements.

**Comment by tkelman** [comment]
I think we'd need to inform Julia's gc that the parent object of an xml element contains a reference to it. I'm not sure how libxml2 handles freeing elements that still have valid references to them. Or if freeing an element also frees all of its children? If there's a memory-management section of libxml2's documentation, I think it would be worth looking into to find out what the proper way of handling this is.

**Comment by robertfeldt** [comment]
The issue seems quite involved. I've looked around and I think this is the most illuminating description of some of the problems and a possible solution:

http://www.omegahat.org/RSXML/MemoryManagement.html

Not sure if its relevant for Julia's context though. For now, just updating the docs so that it is clear for users what to expect is at least a first step. I'll see if I can do a PR on that aspect.

**Comment by robertfeldt** [comment]
This is also illuminating and indicates that xmlUnlinkNode followed by xmlFreeNode might do the trick:

https://mail.gnome.org/archives/xml/2011-April/msg00051.html

**Comment by robertfeldt** [comment]
I experimented a little on my branch freexmlnode. This file shows example code that crashes every time on my machine when I've been trying to add the calls to xmlFreeNode and xmlUnlinkNode:

https://github.com/robertfeldt/LightXML.jl/blob/freexmlnode/spikes/mem_leak_when_creating_many_xmlelems.jl

and here is a typical run with the error:

```julia
feldt:~/dev/forked/LightXML.jl$ julia -L src/LightXML.jl     spikes/mem_leak_when_creating_many_xmlelems.jl
julia(63878,0x7fff7924c300) malloc: *** error for object 0x7f92a0490019: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
/Users/feldt/feldt/bin/julia: line 2: 63878 Abort trap: 6           /Applications/Julia-0.3.5.app/Contents/Resources/julia/bin/julia "$@"
```

**Comment by robertfeldt** [comment]
Here is a pic showing the >13GB of mem retained by the Julia process when running the mem leak spike:

![screen shot 2015-01-28 at 08 25 23](https://cloud.githubusercontent.com/assets/3980/5933900/611b0c64-a6c7-11e4-9ebb-17e60a782cfa.png)

You can see that Max OSX has started compressing memory since it can see it is no longer referenced. But it does not release it.

**Comment by robertfeldt** [comment]
Clarification: The screenshot above was when running against the current LightXML.jl package without any of my changes to try to release memory. When I try calling xmlFreeNode etc I get the malloc error above. So all this is just to try to document the problem and show how others can replicate it; I do not yet have a solution.

**Comment by robertfeldt** [comment]
When I run the spike with command line arguments ```xdoc free``` it never uses more than 200MB on my machine before the mem is gc'd. So a way to avoid the mem leak problems, for now, is to set a finalizer on XMLDocuments directly when you create them. Example code:

```julia
xdoc = XMLDocument()
xmlelem = #... create your xml element tree...
set_root(xdoc, xlmelem)
finalizer(xdoc, free) # This ensures the mem used in the doc and any of its child elements & nodes is GC'd
# use your xdoc...
```

**Comment by freeboson** [comment]
@tkelman should this be closed? I think PR #20 fixed this.
